# Segmenting and Clustering Neighborhoods in Toronto

## Introduction

In this notebook we will scrape the following Wikipedia page ([List of postal codes of Canada: M](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M)) in order to obtain the data that is in the table of postal codes and transform it into a pandas dataframe.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item1">Get the Datasets using BeautifulSoup and pandas</a>

2. <a href="#item2">Explore Neighborhoods using Folium and Foursquare</a>

3. <a href="#item3">Analyze Each Neighborhood</a>

4. <a href="#item4">Cluster Neighborhoods</a>

5. <a href="#item5">Examine Clusters</a>    
</font>
</div>

## 1. Get the Datasets using BeautifulSoup and pandas

We will begin by importing the **BeautifulSoup** package.

In [19]:
from bs4 import BeautifulSoup

We also need to import the **requests** library, which allows to send organic, grass-fed HTTP/1.1 requests, without the need for manual labor. We take the link of the website through which we are going to scrape the data and assign it to variable named wiki_url.

In [20]:
import requests
wiki_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
soup = BeautifulSoup(wiki_url.content,'lxml')
#print(soup.prettify())

We can uncomment the last line in order to take a look at the whole HTML script.

Now, we need to find the class ‘wikitable sortable’ in the HTML script and assign it to the variable `my_table`. Let's take a look to the result.

In [21]:
my_table = soup.find('table',{'class':'wikitable sortable'})
#my_table

We can deduce every `<tr>`...`<\tr>` section corresponds to a row. We can use the `find_all()` method in order to find all the `<tr>` tags in the document and get a set. Now let's see what is the length of this set.

In [22]:
len(my_table.find_all("tr"))

290

The lenght of the set is 290, which means our table contains 290 rows. Now let's import the numpy library to create an empty array of 290 rows and 3 columns, as the table shown in the wiki page.

In [23]:
import numpy as np
matrix = np.empty((290, 3), dtype=object)

Now, we will fill our array with the values of the table using a for loop. Note that we use the `stripped_strings` generator.
When there’s more than one thing inside a tag (as it is our case), you can still look at just the strings using the `.strings` generator. Since these strings tend to have a lot of extra whitespace, you can remove it by using the `.stripped_strings` generator.

In [24]:
for i, val in enumerate(my_table.find_all("tr")):
    for j,string in enumerate(val.stripped_strings):
        matrix[i][j]=string
#matrix

But we wish to keep only the rows that have an assigned borough. We will eliminate the rows where the borough is Not assigned.

In [25]:
matrix2=matrix.transpose()
indices = [i for (i,v) in enumerate(matrix2[1]) if v=='Not assigned']
matrix2 = np.delete(matrix2, indices, 1)
matrix2.shape
for (i,v) in enumerate(matrix2[2]):
    if v=='Not assigned':
        matrix2[2][i]=matrix2[1][i]
# to check if we eliminate the desired rows succesfully the next statement should give 'False' as a result
print('Not assigned' in matrix2[1])
print(matrix2.shape)

False
(3, 213)


We eliminated the unassigned boroughs and we are left with 213 rows. Let's convert it to a **pandas** dataframe.

In [26]:
import pandas as pd
matrix3=matrix2.transpose()
df_torpc=pd.DataFrame(data=matrix3[1:,0:],
                       columns=matrix3[0,0:])
df_torpc.head()

Postcode           Borough     Neighbourhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M5A  Downtown Toronto       Regent Park
4      M6A        North York  Lawrence Heights

More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma.

In [27]:
result = df_torpc.groupby(by=['Postcode','Borough'],sort=False).agg( ', '.join)
result.reset_index(inplace=True)
result.head()

Postcode           Borough                     Neighbourhood
0      M3A        North York                         Parkwoods
1      M4A        North York                  Victoria Village
2      M5A  Downtown Toronto         Harbourfront, Regent Park
3      M6A        North York  Lawrence Heights, Lawrence Manor
4      M7A      Queen's Park                      Queen's Park

In [28]:
result.shape

(103, 3)

Now that we have have built a dataframe of the postal codes along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

I tried to use the Geocoder Python package (https://geocoder.readthedocs.io/index.html) but it didn't work, so I will use the csv file provided here: http://cocl.us/Geospatial_data

In [29]:
import pandas as pd
df_latlng = pd.read_csv('http://cocl.us/Geospatial_data')
df_latlng.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Rename "Postcode" to "Postal Code" and "Neighbourhood" to "Neighborhood"

In [30]:
result.rename(columns={'Postcode': 'Postal Code','Neighbourhood': 'Neighborhood'}, inplace=True)
result.head()

Postal Code           Borough                      Neighborhood
0         M3A        North York                         Parkwoods
1         M4A        North York                  Victoria Village
2         M5A  Downtown Toronto         Harbourfront, Regent Park
3         M6A        North York  Lawrence Heights, Lawrence Manor
4         M7A      Queen's Park                      Queen's Park

Merge both dataframes using the key 'Postal Code'

In [31]:
df_neighb=pd.merge(result, df_latlng, on='Postal Code', how='inner')
df_neighb.head(10)

Postal Code           Borough                      Neighborhood   Latitude  \
0         M3A        North York                         Parkwoods  43.753259   
1         M4A        North York                  Victoria Village  43.725882   
2         M5A  Downtown Toronto         Harbourfront, Regent Park  43.654260   
3         M6A        North York  Lawrence Heights, Lawrence Manor  43.718518   
4         M7A      Queen's Park                      Queen's Park  43.662301   
5         M9A         Etobicoke                  Islington Avenue  43.667856   
6         M1B       Scarborough                    Rouge, Malvern  43.806686   
7         M3B        North York                   Don Mills North  43.745906   
8         M4B         East York   Woodbine Gardens, Parkview Hill  43.706397   
9         M5B  Downtown Toronto          Ryerson, Garden District  43.657162   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494  
5 -79.532242  
6 -79.194353  
7 -79.352188  
8 -79.309937  
9 -79.378937

In [32]:
df_neighb.shape

(103, 5)

In [33]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_neighb['Borough'].unique()),
        df_neighb.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighborhoods.


## 2. Exploring the neighborhoods using Folium and the Foursquare API

Now we will use the Foursquare API to explore neighborhoods in Toronto. We will use the *explore* function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters using the k-means clustering algorithm. Finally, we will use the Folium library to visualize the neighborhoods in Toronto and their emerging clusters.

For this section, we will borrow some of the functions already used in the lab "Segmenting and Clustering Neighborhoods in New York City".

Let's download all the dependencies that we will need.

In [34]:
import json # library to handle JSON files

!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

#import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


#### Use the geopy library to get the latitude and longitude values of Toronto.

In [35]:
address = 'Toronto, ON'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

/home/jupyterlab/conda/lib/python3.6/site-packages/ipykernel_launcher.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  This is separate from the ipykernel package so we can avoid doing imports until


The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### Create a map of Toronto with neighbourhoods superimposed on top.

In [36]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_neighb['Latitude'], df_neighb['Longitude'], df_neighb['Borough'], df_neighb['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Let's simplify the above map and work only with the boroughs that have the word 'Toronto' in their names. So let's slice the original dataframe and create a new dataframe of the Toronto data.

In [37]:
df_toronto=df_neighb[df_neighb['Borough'].str.contains('Toronto')].reset_index(drop=True)
#df_toronto.reset_index(drop=True,inplace=True)
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(df_toronto['Borough'].unique()),
        df_toronto.shape[0]
    )
)
df_toronto.shape

The dataframe has 4 boroughs and 38 neighborhoods.


(38, 5)

Let's visualize again.

In [38]:
# create map of Toronto using latitude and longitude values
map_toronto_rd = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto_rd)  
    
map_toronto_rd

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [39]:
CLIENT_ID = 'NZHMV3O4FE4Z15HFB54BPMTMJ1B1VVKZYX05RWZFIT4DXOWZ' # your Foursquare ID
CLIENT_SECRET = 'FREZIQU3ILKVDTMHALFNU3MMOY2EDJJWWNXS3GZHA4Q4QVAP' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: NZHMV3O4FE4Z15HFB54BPMTMJ1B1VVKZYX05RWZFIT4DXOWZ
CLIENT_SECRET:FREZIQU3ILKVDTMHALFNU3MMOY2EDJJWWNXS3GZHA4Q4QVAP


#### Define function to get the top 100 venues of each neighborhood

This function will get the top 100 venues of each neighborhood using the latitude and longitude coordinates within a radius of 500 meters. The function also structures the json result into a *pandas* dataframe.

In [76]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighbourhood and create a new dataframe called *toronto_venues*.

In [77]:
toronto_venues = getNearbyVenues(names=df_toronto['Neighborhood'],
                                 latitudes=df_toronto['Latitude'],
                                 longitudes=df_toronto['Longitude']
                                )

Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Adelaide, King, Richmond
Dovercourt Village, Dufferin
Harbourfront East, Toronto Islands, Union Station
Little Portugal, Trinity
The Danforth West, Riverdale
Design Exchange, Toronto Dominion Centre
Brockton, Exhibition Place, Parkdale Village
The Beaches West, India Bazaar
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North, Forest Hill West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
Harbord, University of Toronto
Runnymede, Swansea
Moore Park, Summerhill East
Chinatown, Grange Park, Kensington Market
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Rosedale
Stn A PO Boxes 25 The Esplanade
Cabbagetown, St. James Town
Fir

#### Let's check the size of the resulting dataframe

In [78]:
print(toronto_venues.shape)
toronto_venues.head()

(1700, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Harbourfront, Regent Park               43.65426              -79.360636   
1  Harbourfront, Regent Park               43.65426              -79.360636   
2  Harbourfront, Regent Park               43.65426              -79.360636   
3  Harbourfront, Regent Park               43.65426              -79.360636   
4  Harbourfront, Regent Park               43.65426              -79.360636   

                                             Venue  Venue Latitude  \
0                                 Roselle Desserts       43.653447   
1                                    Tandem Coffee       43.653559   
2  Toronto Cooper Koo Family Cherry St YMCA Centre       43.653191   
3                               Morning Glory Cafe       43.653947   
4                              Body Blitz Spa East       43.654735   

   Venue Longitude        Venue Category  
0       -79.362017                Bakery  
1       -79.361809           Coffee Shop  
2       -79.357947  Gym / Fitness Center  
3       -79.361149        Breakfast Spot  
4       -79.359874                   Spa

Let's check how many venues were returned for each neighborhood

In [79]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            54   
Brockton, Exhibition Place, Parkdale Village                           20   
Business Reply Mail Processing Centre 969 Eastern                      17   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            47   
Central Bay Street                                                     85   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   88   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                        7   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     14   
Design Exchange, Toronto Dominion Centre                              100   
Dovercourt Village, Dufferin                                           21   
First Canadian Place, Underground city                                100   
Forest Hill North, Forest Hill West                                     5   
Harbord, University of Toronto                                         36   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              48   
High Park, The Junction South                                          23   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               66   
Moore Park, Summerhill East                                             2   
North Toronto West                                                     19   
Parkdale, Roncesvalles                                                 16   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     35   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        94   
Studio District                                                        40   
The Annex, North Midtown, Yorkville                                    24   
The Beaches                                                             4   
The Beaches West, India Bazaar                                         19   
The Danforth West, Riverdale                                           42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             54   
Brockton, Exhibition Place, Parkdale Village                            20   
Business Reply Mail Processing Centre 969 Eastern                       17   
CN Tower, Bathurst Quay, Island airport, Harbou...                      14   
Cabbagetown, St. James Town                                             47   
Central Bay Street                                                      85   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                16   
Church and Wellesley                                                    88   
Commerce Court, Victoria Hotel     

#### Let's find out how many unique categories can be curated from all the returned venues

In [80]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 238 uniques categories.


## 3. Analyze Each Neighborhood

In [81]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Adult Boutique  Afghan Restaurant  Airport  \
0            0               0                  0        0   
1            0               0                  0        0   
2            0               0                  0        0   
3            0               0                  0        0   
4            0               0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   

   Arepa Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                 0            0           0                    0   
1                 0            0           0                    0   
2                 0            0           0                    0   
3                 0            0           0                    0   
4                 0            0           0                    0   

   Asian Restaurant  Astrologer  Athletics & Sports  Auto Workshop  BBQ Joint  \
0                 0           0                   0              0          0   
1                 0           0                   0              0          0   
2                 0           0                   0              0          0   
3                 0           0                   0              0          0   
4                 0           0                   0              0          0   

   Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0           0           0       1     0    0                 0   
1           0           0       0     0    0                 0   
2           0           0       0     0    0                 0   
3           0           0       0     0    0                 0   
4           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Beer Bar  Beer Store  Belgian Restaurant  \
0                   0      0         0           0                   0   
1                   0      0         0           0                   0   
2                   0      0         0           0                   0   
3                   0      0         0           0                   0   
4                   0      0         0           0                   0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               1        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Bus Stop  Business Service  Butcher  Café  \
0              0         0         0                 0        0     0   
1              0         0         0        

And let's examine the new dataframe size.

In [82]:
toronto_onehot.shape

(1700, 238)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [83]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                            Adelaide, King, Richmond     0.000000   
1                                         Berczy Park     0.000000   
2        Brockton, Exhibition Place, Parkdale Village     0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.000000   
4   CN Tower, Bathurst Quay, Island airport, Harbo...     0.000000   
5                         Cabbagetown, St. James Town     0.000000   
6                                  Central Bay Street     0.011765   
7           Chinatown, Grange Park, Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.011364   
10                     Commerce Court, Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...     0.000000   
14           Design Exchange, Toronto Dominion Centre     0.000000   
15                       Dovercourt Village, Dufferin     0.000000   
16             First Canadian Place, Underground city     0.000000   
17                Forest Hill North, Forest Hill West     0.000000   
18                     Harbord, University of Toronto     0.027778   
19  Harbourfront East, Toronto Islands, Union Station     0.000000   
20                          Harbourfront, Regent Park     0.000000   
21                      High Park, The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                           Little Portugal, Trinity     0.015152   
24                        Moore Park, Summerhill East     0.000000   
25                                 North Toronto West     0.052632   
26                             Parkdale, Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                 Runnymede, Swansea     0.000000   
30                           Ryerson, Garden District     0.000000   
31                                     St. James Town     0.010000   
32                    Stn A PO Boxes 25 The Esplanade     0.010638   
33                                    Studio District     0.050000   
34                The Annex, North Midtown, Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                     The Beaches West, India Bazaar     0.000000   
37                       The Danforth West, Riverdale     0.023810   

    Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0         0.000000           0.000000  0.000000            0.000000   
1         0.000000           0.000000  0.000000            0.000000   
2         0.000000           0.000000  0.000000            0.000000   
3         0.000000           0.000000  0.000000            0.000000   
4         0.000000           0.000000  0.071429            0.071429   
5         0.000000           0.000000  0.000000            0.000000   
6         0.000000           0.000000  0.000000            0.000000   
7         0.000000           0.000000  0.000000            0.000000   
8         0.000000           0.000000  0.000000            0.000000   
9         0.011364           0.011364  0.000000            0.000000   
10        0.000000           0.000000  0.000000            0.000000   
11        0.000000           0.000000  0.000000            0.000000   
12        0.000000           0.000000  0.000000            0.000000   
13        0.000000           0.000000  0.000000            0.000000   
14        0.000000           0.000000  0.000000            0.000000   
15        0.000000           0.000000  0.000000            0.000000   
16        0.000000           0.000000  0.000000            0.000000   
17        0.000000           0.0

#### Let's confirm the new size

In [84]:
toronto_grouped.shape

(38, 238)

#### Let's print each neighborhood along with the top 5 most common venues

In [85]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
                 venue  freq
0          Coffee Shop  0.06
1                 Café  0.05
2  American Restaurant  0.04
3      Thai Restaurant  0.04
4           Steakhouse  0.04


----Berczy Park----
          venue  freq
0   Coffee Shop  0.07
1  Cocktail Bar  0.06
2    Restaurant  0.06
3        Bakery  0.04
4           Pub  0.04


----Brockton, Exhibition Place, Parkdale Village----
                    venue  freq
0             Coffee Shop  0.10
1          Breakfast Spot  0.10
2                    Café  0.10
3  Furniture / Home Store  0.05
4   Performing Arts Venue  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.12
1          Smoke Shop  0.06
2                 Spa  0.06
3       Auto Workshop  0.06
4             Butcher  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0    Airport Lounge  

#### Let's put that into a *pandas* dataframe

This is a function to sort the venues in descending order.

In [86]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [87]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue  \
0                            Adelaide, King, Richmond           Coffee Shop   
1                                         Berczy Park           Coffee Shop   
2        Brockton, Exhibition Place, Parkdale Village           Coffee Shop   
3   Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4   CN Tower, Bathurst Quay, Island airport, Harbo...        Airport Lounge   
5                         Cabbagetown, St. James Town           Coffee Shop   
6                                  Central Bay Street           Coffee Shop   
7           Chinatown, Grange Park, Kensington Market                   Bar   
8                                            Christie         Grocery Store   
9                                Church and Wellesley   Japanese Restaurant   
10                     Commerce Court, Victoria Hotel           Coffee Shop   
11                                         Davisville        Sandwich Place   
12                                   Davisville North                   Gym   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...           Coffee Shop   
14           Design Exchange, Toronto Dominion Centre           Coffee Shop   
15                       Dovercourt Village, Dufferin                Bakery   
16             First Canadian Place, Underground city           Coffee Shop   
17                Forest Hill North, Forest Hill West              Bus Line   
18                     Harbord, University of Toronto                  Café   
19  Harbourfront East, Toronto Islands, Union Station           Coffee Shop   
20                          Harbourfront, Regent Park           Coffee Shop   
21                      High Park, The Junction South                  Café   
22                                      Lawrence Park              Bus Line   
23                           Little Portugal, Trinity                   Bar   
24                        Moore Park, Summerhill East            Restaurant   
25                                 North Toronto West        Clothing Store   
26                             Parkdale, Roncesvalles        Breakfast Spot   
27                                           Rosedale                  Park   
28                                           Roselawn                Garden   
29                                 Runnymede, Swansea           Coffee Shop   
30                           Ryerson, Garden District           Coffee Shop   
31                                     St. James Town           Coffee Shop   
32                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
33                                    Studio District                  Café   
34                The Annex, North Midtown, Yorkville           Coffee Shop   
35                                        The Beaches           Coffee Shop   
36                     The Beaches West, India Bazaar                  Park   
37                       The Danforth West, Riverdale      Greek Restaurant   

   2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0                   Café                     Steakhouse   American Restaurant   
1             Restaurant                   Cocktail Bar    Seafood Restaurant   
2                   Café                 Breakfast Spot             Pet Store   
3             Smoke Shop                  Burrito Place         Auto Workshop   
4        Airport Service               Airport Terminal       Harbor / Marina   
5             Restaurant                       Pharmacy                Bakery   
6                   Café             Italian Restaurant          Burger Joint   
7                   Café  Vegetarian / Vegan Restaurant           Coffee Shop   
8                   Café                           Park             Nightclub   
9       Sushi Restaurant                    Coffee Shop               Gay Bar   
10                  Café                          Hotel            R

## 4. Cluster Neighborhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [88]:
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([3, 3, 3, 3, 3, 3, 3, 3, 3, 3], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [89]:
toronto_merged = df_toronto

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough               Neighborhood   Latitude  \
0         M5A  Downtown Toronto  Harbourfront, Regent Park  43.654260   
1         M5B  Downtown Toronto   Ryerson, Garden District  43.657162   
2         M5C  Downtown Toronto             St. James Town  43.651494   
3         M4E      East Toronto                The Beaches  43.676357   
4         M5E  Downtown Toronto                Berczy Park  43.644771   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.360636               3           Coffee Shop                Bakery   
1 -79.378937               3           Coffee Shop        Clothing Store   
2 -79.375418               3           Coffee Shop            Restaurant   
3 -79.293031               3           Coffee Shop                   Pub   
4 -79.373306               3           Coffee Shop            Restaurant   

  3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
0                  Park                  Café                        Pub   
1                  Café        Cosmetics Shop  Middle Eastern Restaurant   
2                  Café                 Hotel                       Park   
3            Astrologer    Dim Sum Restaurant         Falafel Restaurant   
4          Cocktail Bar    Seafood Restaurant                Cheese Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Restaurant    Mexican Restaurant               Theater   
1            Restaurant                   Bar   Japanese Restaurant   
2                Bakery             Gastropub    Italian Restaurant   
3           Event Space  Ethiopian Restaurant     Electronics Store   
4            Steakhouse    Italian Restaurant        Farmers Market   

         9th Most Common Venue 10th Most Common Venue  
0               Breakfast Spot             Shoe Store  
1           Italian Restaurant       Ramen Restaurant  
2               Breakfast Spot         Clothing Store  
3  Eastern European Restaurant    Dumpling Restaurant  
4                       Bakery                   Café

Finally, let's visualize the resulting clusters

In [90]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

Now, we can examine each cluster and determine the discriminating venue categories that distinguish each cluster.

#### Cluster 1

In [91]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
12  Downtown Toronto               0           Coffee Shop   
17   Central Toronto               0              Bus Line   
22   Central Toronto               0        Clothing Store   
36  Downtown Toronto               0   Japanese Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12                  Café                 Hotel   American Restaurant   
17                  Park           Swim School    Dim Sum Restaurant   
22           Coffee Shop   Sporting Goods Shop            Bagel Shop   
36      Sushi Restaurant           Coffee Shop               Gay Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
12            Restaurant    Italian Restaurant             Gastropub   
17          Dessert Shop           Event Space  Ethiopian Restaurant   
22  Fast Food Restaurant                 Diner    Mexican Restaurant   
36            Restaurant          Burger Joint                   Pub   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
12         Deli / Bodega                          Gym             Sports Bar  
17     Electronics Store  Eastern European Restaurant    Dumpling Restaurant  
22          Dessert Shop                         Park     Chinese Restaurant  
36                  Café                    Gastropub   Fast Food Restaurant

#### Cluster 2

In [92]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
28  Central Toronto               1            Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28            Playground      Department Store           Event Space   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
28  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28   Dumpling Restaurant            Donut Shop       Doner Restaurant

#### Cluster 3

In [96]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
24  West Toronto               2        Breakfast Spot             Gift Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
24             Bookstore            Restaurant               Dog Run   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
24          Burger Joint         Movie Theater          Dessert Shop   

   9th Most Common Venue 10th Most Common Venue  
24                  Bank                    Bar

#### Cluster 4

In [97]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               3           Coffee Shop   
1   Downtown Toronto               3           Coffee Shop   
2   Downtown Toronto               3           Coffee Shop   
3       East Toronto               3           Coffee Shop   
4   Downtown Toronto               3           Coffee Shop   
5   Downtown Toronto               3           Coffee Shop   
6   Downtown Toronto               3         Grocery Store   
7   Downtown Toronto               3           Coffee Shop   
8       West Toronto               3                Bakery   
9   Downtown Toronto               3           Coffee Shop   
10      West Toronto               3                   Bar   
11      East Toronto               3      Greek Restaurant   
13      West Toronto               3           Coffee Shop   
14      East Toronto               3                  Park   
15  Downtown Toronto               3           Coffee Shop   
16      East Toronto               3                  Café   
18   Central Toronto               3                Garden   
19   Central Toronto               3                   Gym   
20   Central Toronto               3              Bus Line   
21      West Toronto               3                  Café   
23   Central Toronto               3           Coffee Shop   
25   Central Toronto               3        Sandwich Place   
26  Downtown Toronto               3                  Café   
29  Downtown Toronto               3                   Bar   
30   Central Toronto               3           Coffee Shop   
31  Downtown Toronto               3        Airport Lounge   
32  Downtown Toronto               3                  Park   
33  Downtown Toronto               3           Coffee Shop   
34  Downtown Toronto               3           Coffee Shop   
35  Downtown Toronto               3           Coffee Shop   
37      East Toronto               3    Light Rail Station   

   2nd Most Common Venue          3rd Most Common Venue 4th Most Common Venue  \
0                 Bakery                           Park                  Café   
1         Clothing Store                           Café        Cosmetics Shop   
2             Restaurant                           Café                 Hotel   
3                    Pub                     Astrologer    Dim Sum Restaurant   
4             Restaurant                   Cocktail Bar    Seafood Restaurant   
5                   Café             Italian Restaurant          Burger Joint   
6                   Café                           Park             Nightclub   
7                   Café                     Steakhouse   American Restaurant   
8            Supermarket                 Discount Store              Pharmacy   
9                  Hotel                       Aquarium                  Café   
10           Men's Store               Asian Restaurant           Coffee Shop   
11           Coffee Shop                 Ice Cream Shop             Bookstore   
13                  Café                 Breakfast Spot             Pet Store   
14      Sushi Restaurant                 Ice Cream Shop        Sandwich Place   
15                  Café                          Hotel            Restaurant   
16           Coffee Shop             Italian Restaurant                Bakery   
18          Home Service                   Dessert Shop    Falafel Restaurant   
19                  Park                 Breakfast Spot     Food & Drink Shop   
20         Jewelry Store                           Park      Sushi Restaurant   
21    Mexican Restaurant                 Sandwich Place             Speakeasy   
23                  Café                 Sandwich Place           Pizza Place   
25          Dessert Shop                    Pizza Place      Sushi Restaurant   
26                Bakery                            Bar   Japanese Restaurant   
29                  Café  Vegetarian / Vegan Restaurant           Coffee Shop   
30 

#### Cluster 5

In [98]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
27  West Toronto               4           Coffee Shop           Pizza Place   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
27                  Café      Sushi Restaurant    Italian Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
27             Bookstore   Indie Movie Theater     Fish & Chips Shop   

   9th Most Common Venue 10th Most Common Venue  
27                   Bar          Smoothie Shop